File: go/core/netpolicy/netpolicy.go
// Package netpolicy owns the process-wide network policy marker and the
// http.RoundTripper that enforces it.
//
// The --offline global (cli-parity-guide, "Global Flags") promises that
// network access is disabled. Stamping a context value alone cannot keep
// that promise: it is advisory, so any caller that forgets to consult it
// still reaches the wire. Guard closes that gap by refusing the request
// inside the transport, beneath every http.Client, where no caller can
// route around it.
//
// The marker lives here rather than in go/console/cli because cli imports
// go/core/*; transports enforcing the policy would otherwise import cli
// back and cycle. cli re-exports WithOffline/IsOffline as forwarders so
// the existing call sites keep working.
//
// Loopback is deliberately exempt. --offline means "do not talk to the
// network", not "do not talk to myself": a local `kit serve` peer, a dev
// backend on 127.0.0.1 and unix sockets stay reachable so offline
// workflows remain usable.
//
// # Scope
//
// Enforcement has two seams, because Go offers no single one.
//
// Guard sits in the http.RoundTripper chain, and Install puts it under
// http.DefaultTransport, so every client that does not set its own
// Transport is covered without a per-site change. That is all HTTP and
// HTTPS through net/http, and with it ConnectRPC, which rides an
// ordinary *http.Client.
//
// GuardDial is the seam for everything net/http does not own. Go has no
// hook beneath net.Dial: a library that opens its own socket cannot be
// intercepted from outside, so it is covered only once its dial function
// is routed through GuardDial. Libraries that accept an injected dialer
// — net/smtp via net.Dialer, coder/websocket via DialOptions.HTTPClient,
// go-sql-driver/mysql via Config.DialFunc or RegisterDialContext, gRPC
// via WithContextDialer, crypto/tls via tls.Client over a dialed conn —
// can therefore all be brought under the policy. CheckDial exposes the
// same decision for hooks that are not shaped like a dialer.
//
// # kit's own egress
//
// Every egress path kit itself owns is now routed through one of the two
// seams:
//
//   - SMTP (runtime/notify/sinks/email) dials through GuardDial. The
//     mailer still accepts a *net.Dialer, which cannot itself be made
//     policy-aware; the guard wraps its DialContext at the call site.
//   - WebSocket (transport/api/client and runtime/bus) hands
//     coder/websocket an *http.Client whose transport is Guard-wrapped.
//     Naming it explicitly rather than leaning on http.DefaultClient
//     keeps enforcement independent of whether Install has run. The bus
//     reconnect loop starts from its own root context and replays the
//     policy its last Connect saw, so a dropped peer is not re-dialed
//     on an offline run.
//   - TiDB/MySQL (storage/kv/tidb) sets Config.DialFunc to a guarded
//     dialer. The driver invokes it with the context of the query that
//     triggered the connection, so the marker survives sql.Open's
//     laziness and the refusal lands on first use.
//
// One gap remains inside kit, bounded by an API rather than by Go:
// kv.Open and kv.Opener carry no context, so a Store opened through the
// backend registry cannot police its own open-time ping. The guarded
// dial hook is installed regardless, so every query through that Store
// is covered; only the initial connect escapes. Callers with a context
// should use tidb.NewContext.
//
// What remains uncovered, and cannot be covered from this package:
//
//   - A dependency that calls net.Dial itself and exposes no dialer
//     hook. Nothing in Go can intercept it; the only fixes are upstream
//     or a sandbox outside the process.
//   - A client holding a *net.Dialer rather than a dial function. Only
//     DialContext carries a context, and DialContext is a method, not a
//     replaceable field, so a *net.Dialer cannot be made policy-aware.
//     Such a call site must call GuardDial's result directly.
//   - A caller that captured http.DefaultTransport before Install ran,
//     or that sets an explicit Transport without wrapping it in Guard.
//
// For anything in that list --offline stays advisory and the call site
// must consult IsOffline itself.
package netpolicy

import (
	"context"
	"errors"
	"fmt"
	"net"
	"net/http"
)

// ErrOffline is returned by Guard when a request is attempted on a
// context marked offline. Callers match it with errors.Is; it is wrapped
// in a *url.Error by net/http, which errors.Is unwraps.
var ErrOffline = errors.New("network disabled by --offline")

type offlineCtxKey struct{}

// WithOffline returns a context marked as offline. Passing false returns
// ctx unchanged so an untagged context stays clean.
func WithOffline(ctx context.Context, offline bool) context.Context {
	if !offline {
		return ctx
	}
	return context.WithValue(ctx, offlineCtxKey{}, true)
}

// IsOffline reports whether ctx carries the offline marker. nil-safe.
func IsOffline(ctx context.Context) bool {
	if ctx == nil {
		return false
	}
	on, _ := ctx.Value(offlineCtxKey{}).(bool)
	return on
}

// isLoopback reports whether host names a loopback address. Hosts that
// are not literal IPs (DNS names) are treated as remote: resolving them
// would itself be network access.
func isLoopback(host string) bool {
	h, _, err := net.SplitHostPort(host)
	if err != nil {
		h = host
	}
	if h == "localhost" {
		return true
	}
	ip := net.ParseIP(h)
	return ip != nil && ip.IsLoopback()
}

// guard is the RoundTripper returned by Guard.
type guard struct{ base http.RoundTripper }

// Guard wraps base so requests on an offline-marked context fail with
// ErrOffline instead of reaching the network. A nil base falls back to
// http.DefaultTransport. Wrapping is idempotent: guarding an already
// guarded transport returns it unchanged.
func Guard(base http.RoundTripper) http.RoundTripper {
	if base == nil {
		base = http.DefaultTransport
	}
	if g, ok := base.(*guard); ok {
		return g
	}
	return &guard{base: base}
}

// RoundTrip refuses the request when its context is marked offline and
// the destination is not loopback.
func (g *guard) RoundTrip(req *http.Request) (*http.Response, error) {
	if IsOffline(req.Context()) && !isLoopback(req.URL.Host) {
		return nil, fmt.Errorf("%s %s: %w", req.Method, req.URL.Redacted(), ErrOffline)
	}
	return g.base.RoundTrip(req)
}

// Install wraps http.DefaultTransport with Guard, so every client that
// does not set an explicit Transport — the common case across kit and
// adopter code — enforces the policy without a per-site change.
//
// It is idempotent and safe to call more than once. Call it once during
// process start-up (cli.New does this) and never concurrently with
// in-flight requests: it mutates a process-global.
//
// Clients that DO set their own Transport must wrap it themselves with
// Guard; Install cannot reach them.
func Install() {
	http.DefaultTransport = Guard(http.DefaultTransport)
}

// # Observability carve-out
//
// --offline means "do not talk to the network on the user's behalf". It
// does not silence diagnostics: telemetry, and any future remote-logging
// or crash-reporting sink, are logging-class egress and stay exempt, the
// same way a remote syslog target would not be muted by an offline flag.
// Consent and the telemetry mode already govern whether those emit at
// all; --offline is not a second consent gate.
//
// Sinks opt out by building their client with ObservabilityTransport
// rather than relying on http.DefaultTransport, which Install guards.
// The exemption is therefore explicit at the call site and survives a
// sink changing which HTTP library it uses.

// ObservabilityTransport returns a transport that ignores the offline
// marker, for logging-class sinks (telemetry, remote logging, crash
// reporting). base may be nil, in which case the unguarded default is
// used.
//
// If base is already guarded, the guard is stripped: the caller is
// asserting this client carries diagnostics, not user-initiated traffic.
// Never use it for anything the user asked for — that is exactly the
// traffic --offline exists to stop.
func ObservabilityTransport(base http.RoundTripper) http.RoundTripper {
	if g, ok := base.(*guard); ok {
		return g.base
	}
	if base != nil {
		return base
	}
	if g, ok := http.DefaultTransport.(*guard); ok {
		return g.base
	}
	return http.DefaultTransport
}

File: sdk/experimental/rs/src/netpolicy.rs
//! Network policy marker and the client that enforces it.
//!
//! The `--offline` global (cli-parity-guide, "Global Flags") promises
//! that network access is disabled. Carrying that promise as a plain
//! flag value cannot keep it: it is advisory, so any caller that forgets
//! to consult it still reaches the wire. [`GuardedClient`] closes that
//! gap by refusing the request at the point where it is sent, beneath
//! every caller, where none can route around it.
//!
//! # Why this shape
//!
//! The Go side installs an `http.RoundTripper` over the process-global
//! `http.DefaultTransport`, so a bare `&http.Client{}` inherits the
//! policy without a per-site change. Rust has no such process-global:
//! `reqwest::Client::new()` builds an independent stack every time, and
//! nothing reqwest 0.12 exposes lets a policy see a request before it is
//! sent. Its `connector_layer` hook receives an opaque `Unnameable` (no
//! destination to inspect) and fires only on new connections, so pooled
//! reuse would slip past; a custom `dns_resolver` never sees IP-literal
//! URLs at all and cannot surface a typed error to the caller.
//!
//! So the chokepoint here is construction rather than transport: this
//! module owns the port's only `reqwest::Client` construction path, and
//! hands back a [`GuardedClient`] that does not expose the inner
//! `reqwest::Client`. Every request necessarily passes through
//! [`GuardedClient::execute`], which is where the policy is applied.
//!
//! # Marker
//!
//! Rust has no ambient request-scoped context, so the marker is an
//! explicit [`NetPolicy`] value threaded at client construction rather
//! than a context value. [`NetPolicy::default`] is online, so a port
//! that has not wired the flag behaves exactly as it did before.
//!
//! Loopback is deliberately exempt. `--offline` means "do not talk to
//! the network", not "do not talk to myself": a local peer, a dev
//! backend on 127.0.0.1 and unix sockets stay reachable so offline
//! workflows remain usable.
//!
//! # Scope
//!
//! [`GuardedClient`] covers HTTP and HTTPS through reqwest — every
//! network client in this port today. It does NOT cover code that opens
//! a socket directly: raw `TcpStream`, SQL drivers, gRPC or raw TLS. For
//! those, `--offline` remains advisory and the call site must consult
//! [`NetPolicy::is_offline`] itself. It also cannot reach a
//! `reqwest::Client` a consumer builds on its own, outside this module —
//! that is the price of Rust having no default transport to install
//! over.

use std::net::IpAddr;

use thiserror::Error;

/// The network policy marker. Threaded at client construction; the
/// analogue of the Go side's context value.
///
/// Default is online, so an unwired port keeps its current behaviour.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub struct NetPolicy {
    offline: bool,
}

impl NetPolicy {
    /// Build a policy from a boolean, mirroring Go's
    /// `WithOffline(ctx, bool)`. This is the one-liner the CLI layer
    /// calls once it registers the `--offline` flag.
    pub fn new(offline: bool) -> Self {
        Self { offline }
    }

    /// A policy that permits network access.
    pub fn online() -> Self {
        Self { offline: false }
    }

    /// A policy that refuses non-loopback network access.
    pub fn offline() -> Self {
        Self { offline: true }
    }

    /// Whether this policy refuses non-loopback network access.
    pub fn is_offline(&self) -> bool {
        self.offline
    }

    /// Whether a request to `url` is permitted under this policy.
    /// Loopback destinations are always permitted.
    pub fn allows(&self, url: &reqwest::Url) -> bool {
        !self.offline || url.host_str().is_some_and(is_loopback)
    }
}

/// The typed refusal, the analogue of Go's `ErrOffline`. Callers match
/// it with [`NetError::as_offline`] or by downcasting.
#[derive(Debug, Clone, Error, PartialEq, Eq)]
#[error("{method} {url}: network disabled by --offline")]
pub struct OfflineError {
    /// HTTP method of the refused request.
    pub method: String,
    /// Destination of the refused request, with any userinfo redacted.
    pub url: String,
}

/// Errors surfaced by [`GuardedClient`]. Refusals are a distinct,
/// matchable variant; everything else is reqwest's own error.
#[derive(Debug, Error)]
pub enum NetError {
    /// The request was refused by the offline policy.
    #[error(transparent)]
    Offline(#[from] OfflineError),
    /// The request was attempted and reqwest failed it.
    #[error(transparent)]
    Transport(#[from] reqwest::Error),
}

impl NetError {
    /// The offline refusal, if this error is one. The analogue of
    /// `errors.Is(err, netpolicy.ErrOffline)`.
    pub fn as_offline(&self) -> Option<&OfflineError> {
        match self {
            NetError::Offline(e) => Some(e),
            NetError::Transport(_) => None,
        }
    }

    /// Whether this error is an offline refusal.
    pub fn is_offline(&self) -> bool {
        self.as_offline().is_some()
    }
}

/// Whether `host` names a loopback address.
///
/// Hosts that are not literal IPs (DNS names) are treated as remote:
/// resolving them would itself be network access. Matches Go's
/// `isLoopback`, including the bare `localhost` special case and the
/// whole 127.0.0.0/8 range.
pub fn is_loopback(host: &str) -> bool {
    if host.eq_ignore_ascii_case("localhost") {
        return true;
    }
    // reqwest::Url::host_str keeps IPv6 literals bracketed; strip them
    // so the address parses.
    let bare = host
        .strip_prefix('[')
        .and_then(|h| h.strip_suffix(']'))
        .unwrap_or(host);
    bare.parse::<IpAddr>().is_ok_and(|ip| ip.is_loopback())
}

/// The port's shared HTTP client: the single construction path through
/// which every reqwest request in this crate is issued.
///
/// It deliberately does not expose its inner `reqwest::Client`. Handing
/// one out would let a caller send a request that never reaches
/// [`GuardedClient::execute`], which is the only place the policy is
/// applied.
#[derive(Debug, Clone)]
pub struct GuardedClient {
    inner: reqwest::Client,
    policy: NetPolicy,
}

impl GuardedClient {
    /// Build a client under `policy` with reqwest's defaults.
    pub fn new(policy: NetPolicy) -> Result<Self, reqwest::Error> {
        Self::build(policy, reqwest::Client::builder())
    }

    /// Build a client under `policy` from a configured
    /// `reqwest::ClientBuilder`, for callers that need timeouts or other
    /// transport settings. The policy is applied regardless of how the
    /// builder was configured.
    pub fn build(
        policy: NetPolicy,
        builder: reqwest::ClientBuilder,
    ) -> Result<Self, reqwest::Error> {
        Ok(Self {
            inner: builder.build()?,
            policy,
        })
    }

    /// The policy this client enforces.
    pub fn policy(&self) -> NetPolicy {
        self.policy
    }

    /// Start a request. The returned builder is reqwest's own, so the
    /// full builder surface stays available; it must be handed back to
    /// [`GuardedClient::send`] (or [`GuardedClient::execute`]) to be
    /// issued.
    pub fn request(&self, method: reqwest::Method, url: impl reqwest::IntoUrl) -> RequestBuilder {
        RequestBuilder {
            client: self.clone(),
            inner: self.inner.request(method, url),
        }
    }

    /// Start a GET request.
    pub fn get(&self, url: impl reqwest::IntoUrl) -> RequestBuilder {
        self.request(reqwest::Method::GET, url)
    }

    /// Start a POST request.
    pub fn post(&self, url: impl reqwest::IntoUrl) -> RequestBuilder {
        self.request(reqwest::Method::POST, url)
    }

    /// Issue `req`, refusing it when the policy forbids the
    /// destination. This is the chokepoint: no request leaves this
    /// crate without passing through here.
    pub async fn execute(&self, req: reqwest::Request) -> Result<reqwest::Response, NetError> {
        if !self.policy.allows(req.url()) {
            let mut url = req.url().clone();
            // Match Go's url.Redacted(): never echo credentials back in
            // an error message.
            if !url.username().is_empty() || url.password().is_some() {
                let _ = url.set_username("xxxxx");
                let _ = url.set_password(Some("xxxxx"));
            }
            return Err(OfflineError {
                method: req.method().to_string(),
                url: url.to_string(),
            }
            .into());
        }
        Ok(self.inner.execute(req).await?)
    }
}

/// A request under construction against a [`GuardedClient`].
///
/// Wraps reqwest's builder so the only way to issue the request is
/// [`RequestBuilder::send`], which routes through the guard.
#[derive(Debug)]
pub struct RequestBuilder {
    client: GuardedClient,
    inner: reqwest::RequestBuilder,
}

impl RequestBuilder {
    /// Set a header. Takes the string forms rather than reqwest's full
    /// generic bound, which would require a direct `http` dependency
    /// this crate does not otherwise carry.
    pub fn header(self, key: &str, value: &str) -> Self {
        Self {
            inner: self.inner.header(key, value),
            ..self
        }
    }

    /// Set a bearer authorization header.
    pub fn bearer_auth<T: std::fmt::Display>(self, token: T) -> Self {
        Self {
            inner: self.inner.bearer_auth(token),
            ..self
        }
    }

    /// Set the request body.
    pub fn body<T: Into<reqwest::Body>>(self, body: T) -> Self {
        Self {
            inner: self.inner.body(body),
            ..self
        }
    }

    /// Serialize `json` as the JSON request body.
    pub fn json<T: serde::Serialize + ?Sized>(self, json: &T) -> Self {
        Self {
            inner: self.inner.json(json),
            ..self
        }
    }

    /// Append serialized `query` to the URL's query string.
    pub fn query<T: serde::Serialize + ?Sized>(self, query: &T) -> Self {
        Self {
            inner: self.inner.query(query),
            ..self
        }
    }

    /// Issue the request through the guard.
    pub async fn send(self) -> Result<reqwest::Response, NetError> {
        let req = self.inner.build()?;
        self.client.execute(req).await
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn policy_allows_loopback_while_offline() {
        let p = NetPolicy::offline();
        assert!(p.allows(&"http://127.0.0.1:8080/x".parse().unwrap()));
        assert!(p.allows(&"http://localhost/x".parse().unwrap()));
        assert!(p.allows(&"http://[::1]:9000/x".parse().unwrap()));
        assert!(!p.allows(&"https://example.com/x".parse().unwrap()));
    }

    #[test]
    fn online_policy_allows_everything() {
        let p = NetPolicy::online();
        assert!(p.allows(&"https://example.com/x".parse().unwrap()));
        assert!(p.allows(&"http://127.0.0.1/x".parse().unwrap()));
    }

    #[test]
    fn offline_error_message_names_the_flag() {
        let e = OfflineError {
            method: "GET".into(),
            url: "https://example.com/x".into(),
        };
        assert_eq!(
            e.to_string(),
            "GET https://example.com/x: network disabled by --offline"
        );
    }

    #[test]
    fn net_error_discriminates() {
        let e: NetError = OfflineError {
            method: "GET".into(),
            url: "https://example.com/x".into(),
        }
        .into();
        assert!(e.is_offline());
        assert_eq!(e.as_offline().unwrap().method, "GET");
    }
}
